File: src/main/java/uk/ac/ebi/tsc/tesk/util/ExecutorCommandWrapper.java
package uk.ac.ebi.tsc.tesk.util;

import org.apache.commons.lang.StringUtils;
import uk.ac.ebi.tsc.tesk.model.TesExecutor;

import java.util.ArrayList;
import java.util.List;
import java.util.StringJoiner;

/**
 * Wraps list of executor's command, so that:
 * - if any of executor's stdin/stdout/stderr params is set, the command runs in shell
 * - stdin, stdout, stderr streams are redirected to paths according to executors params
 * - no shell argument escaping at the moment
 */
public class ExecutorCommandWrapper {
    private final TesExecutor executor;

    public ExecutorCommandWrapper(TesExecutor executor) {
        this.executor = executor;
    }

    public List<String> getCommandsWithStreamRedirects() {
        List<String> result = new ArrayList<>();
        if (StringUtils.isEmpty(executor.getStdin()) && StringUtils.isEmpty(executor.getStdout()) && StringUtils.isEmpty(executor.getStderr())) {
            return executor.getCommand();
        }
        result.add("/bin/sh");
        result.add("-c");
        StringJoiner sj = new StringJoiner(" ");
        executor.getCommand().stream().forEach(sj::add);
        if (!StringUtils.isEmpty(executor.getStdin())) {
            sj.add("<");
            sj.add(executor.getStdin());
        }
        if (!StringUtils.isEmpty(executor.getStdout())) {
            sj.add(">");
            sj.add(executor.getStdout());
        }
        if (!StringUtils.isEmpty(executor.getStderr())) {
            sj.add("2>");
            sj.add(executor.getStderr());
        }
        result.add(sj.toString());
        return result;
    }
}

File: tesk/k8s/converter/executor_command_wrapper.py
"""Wraps list of executor's command.

Such that:
- If any of executor's stdin/stdout/stderr params is set, the command runs in shell
- Each part of the original command (single command/argument) that contained shell
  special chars is surrounded by single quotes, plus single quote inside such string
  are replaced with '"'"' sequence
- `stdin`, `stdout`, `stderr` streams are redirected to paths according to executors
  params
"""

from typing import List

from tesk.api.ga4gh.tes.models import TesExecutor


class ExecutorCommandWrapper:
    """Wraps executor's command.
    
    Attributes:
        executor: executor to wrap
    """

    def __init__(self, executor: TesExecutor):
        """Initialize the wrapper.
        
        Args:
            executor: executor to wrap
        """
        self.executor = executor

    def get_commands_with_stream_redirects(self) -> List[str]:
        """Get command with stream redirects.
        
        Returns:
            List[str]: command to run by exector with stream redirects
        """
        if (
            not self.executor.stdin
            and not self.executor.stdout
            and not self.executor.stderr
        ):
            return self.executor.command

        result = ["/bin/sh", "-c"]

        command_parts = [" ".join(self.executor.command)]

        if self.executor.stdin:
            command_parts.extend(["<", self.executor.stdin])
        if self.executor.stdout:
            command_parts.extend([">", self.executor.stdout])
        if self.executor.stderr:
            command_parts.extend(["2>", self.executor.stderr])

        result.append(" ".join(command_parts))

        return result
